File: python/22_generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []

        def backtrack(s, left, right):
            if len(s) == 2 * n:
                ans.append(s)
                return
            if left < n:
                backtrack(s + '(', left + 1, right)
            if right < n and left > right:
                backtrack(s + ')', left, right + 1)

        backtrack('', 0, 0)

        return ans

File: go/22_generate-parentheses.go
func generateParenthesis(n int) []string {
	var ans []string
	backtrack(&ans, "", 0, 0, n)
	return ans
}

func backtrack(ans *[]string, s string, left int, right int, n int) {
	if len(s) == 2*n {
		*ans = append(*ans, s)
	}
	if left < n {
		backtrack(ans, s+"(", left+1, right, n)
	}
	if left > right {
		backtrack(ans, s+")", left, right+1, n)
	}
}